File: examples/cpp/casbot2_cpp_demo/src/service_mode_demo.cpp
#include <chrono>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_srvs/srv/set_bool.hpp"
#include "crb_ros_msg/srv/get_robot_mode.hpp"
#include "crb_ros_msg/srv/get_robot_state.hpp"
#include "crb_ros_msg/srv/set_robot_mode.hpp"

using namespace std::chrono_literals;

template<typename ServiceT>
bool wait_service(const rclcpp::Client<ServiceT>::SharedPtr &client, const std::string &name, rclcpp::Node &node)
{
  if (!client->wait_for_service(3s)) {
    RCLCPP_ERROR(node.get_logger(), "服务不可用: %s", name.c_str());
    return false;
  }
  return true;
}

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = rclcpp::Node::make_shared("service_mode_demo");

  auto get_mode = node->create_client<crb_ros_msg::srv::GetRobotMode>("get_robot_mode");
  auto set_mode = node->create_client<crb_ros_msg::srv::SetRobotMode>("/set_robot_mode");
  auto get_state = node->create_client<crb_ros_msg::srv::GetRobotState>("get_robot_state_srv_hl");
  auto switch_nav = node->create_client<std_srvs::srv::SetBool>("/motion/switch_nav_mode");
  auto switch_teleop = node->create_client<std_srvs::srv::SetBool>("/switch_teleoperation");
  auto switch_auto = node->create_client<std_srvs::srv::SetBool>("/switch_autonomous");

  if (!wait_service(get_mode, "get_robot_mode", *node) ||
      !wait_service(set_mode, "/set_robot_mode", *node) ||
      !wait_service(get_state, "get_robot_state_srv_hl", *node) ||
      !wait_service(switch_nav, "/motion/switch_nav_mode", *node) ||
      !wait_service(switch_teleop, "/switch_teleoperation", *node) ||
      !wait_service(switch_auto, "/switch_autonomous", *node)) {
    rclcpp::shutdown();
    return 1;
  }

  {
    auto req = std::make_shared<crb_ros_msg::srv::GetRobotMode::Request>();
    auto fut = get_mode->async_send_request(req);
    if (rclcpp::spin_until_future_complete(node, fut, 3s) == rclcpp::FutureReturnCode::SUCCESS) {
      const auto resp = fut.get();
      RCLCPP_INFO(node->get_logger(), "当前模式: mode=%d, mode_name=%s", resp->mode, resp->mode_name.c_str());
    }
  }

  {
    auto req = std::make_shared<crb_ros_msg::srv::GetRobotState::Request>();
    req->start = true;
    auto fut = get_state->async_send_request(req);
    if (rclcpp::spin_until_future_complete(node, fut, 3s) == rclcpp::FutureReturnCode::SUCCESS) {
      const auto resp = fut.get();
      RCLCPP_INFO(node->get_logger(), "机器人状态: state=%u", resp->state);
    }
  }

  {
    auto req = std::make_shared<crb_ros_msg::srv::SetRobotMode::Request>();
    req->mode_name = "WALK";
    auto fut = set_mode->async_send_request(req);
    if (rclcpp::spin_until_future_complete(node, fut, 3s) == rclcpp::FutureReturnCode::SUCCESS) {
      RCLCPP_INFO(node->get_logger(), "set_robot_mode(WALK): %s", fut.get()->success ? "success" : "failed");
    }
  }

  auto call_set_bool = [&node](const std::string &name, const rclcpp::Client<std_srvs::srv::SetBool>::SharedPtr &cli, bool data) {
    auto req = std::make_shared<std_srvs::srv::SetBool::Request>();
    req->data = data;
    auto fut = cli->async_send_request(req);
    if (rclcpp::spin_until_future_complete(node, fut, 3s) == rclcpp::FutureReturnCode::SUCCESS) {
      const auto resp = fut.get();
      RCLCPP_INFO(node->get_logger(), "%s(%s): success=%s, msg=%s",
        name.c_str(), data ? "true" : "false", resp->success ? "true" : "false", resp->message.c_str());
    }
  };

  call_set_bool("/motion/switch_nav_mode", switch_nav, true);
  call_set_bool("/motion/switch_nav_mode", switch_nav, false);
  call_set_bool("/switch_teleoperation", switch_teleop, true);
  call_set_bool("/switch_teleoperation", switch_teleop, false);
  call_set_bool("/switch_autonomous", switch_auto, true);
  call_set_bool("/switch_autonomous", switch_auto, false);

  rclcpp::shutdown();
  return 0;
}

File: examples/python/casbot2_py_demo/casbot2_py_demo/service_mode_demo.py
import rclpy
from rclpy.node import Node
from std_srvs.srv import SetBool

from crb_ros_msg.srv import GetRobotMode, GetRobotState, SetRobotMode


class ServiceModeDemo(Node):
    def __init__(self) -> None:
        super().__init__("service_mode_demo")
        self.get_mode_cli = self.create_client(GetRobotMode, "get_robot_mode")
        self.set_mode_cli = self.create_client(SetRobotMode, "/set_robot_mode")
        self.get_state_cli = self.create_client(GetRobotState, "get_robot_state_srv_hl")
        self.nav_cli = self.create_client(SetBool, "/motion/switch_nav_mode")
        self.teleop_cli = self.create_client(SetBool, "/switch_teleoperation")
        self.auto_cli = self.create_client(SetBool, "/switch_autonomous")

    def call(self, cli, req, timeout_sec=3.0):
        if not cli.wait_for_service(timeout_sec=timeout_sec):
            self.get_logger().error(f"服务不可用: {cli.srv_name}")
            return None
        fut = cli.call_async(req)
        rclpy.spin_until_future_complete(self, fut, timeout_sec=timeout_sec)
        return fut.result()

    def run(self) -> None:
        mode = self.call(self.get_mode_cli, GetRobotMode.Request())
        if mode:
            self.get_logger().info(f"当前模式: mode={mode.mode}, mode_name={mode.mode_name}")

        state_req = GetRobotState.Request()
        state_req.start = True
        state = self.call(self.get_state_cli, state_req)
        if state:
            self.get_logger().info(f"机器人状态: state={state.state}")

        set_req = SetRobotMode.Request()
        set_req.mode_name = "WALK"
        set_resp = self.call(self.set_mode_cli, set_req)
        if set_resp:
            self.get_logger().info(f"set_robot_mode(WALK): {set_resp.success}")

        for name, cli, enable in [
            ("switch_nav_mode", self.nav_cli, True),
            ("switch_nav_mode", self.nav_cli, False),
            ("switch_teleoperation", self.teleop_cli, True),
            ("switch_teleoperation", self.teleop_cli, False),
            ("switch_autonomous", self.auto_cli, True),
            ("switch_autonomous", self.auto_cli, False),
        ]:
            req = SetBool.Request()
            req.data = enable
            resp = self.call(cli, req)
            if resp:
                self.get_logger().info(
                    f"{name}({enable}) => success={resp.success}, msg={resp.message}"
                )


def main() -> None:
    rclpy.init()
    node = ServiceModeDemo()
    node.run()
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()
